---
title: "MICB405 - Differential Gene Analysis"
author: "Carlo von Vegesack"
date: "2024-11-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE, message = FALSE)
```
### Setup for first installation
Only execute if you re-installed R or never installed BiocManager
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
BiocManager::install("pheatmap")
```

### Normal start
Set the working directory as you need it. In my case I set the working directory to a folder called "final_project". Inside this folder is a folder called "count_matrix". "count_matrix" has the ReadsPerGene files. You can see this also in the Load Count Matrix part.
```{r}
# Load necessary libraries
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(RColorBrewer))

# Set the working directory
setwd("~/uni/autumn-24/final_project/")
```
## Load tRNA gene ids
```{r}
# Load tRNA gene IDs
tRNA_gene_ids <- readLines("tRNA_gene_ids.txt")

# Check the first few tRNA gene IDs
head(tRNA_gene_ids)

```
## Load Count Matrix
Loading all ReadsPerGene files from the folder "count_matrix". For the code to run properly it is important to have the ReadsPerGene files named a certain way. I named them treatment_time_replica+normal_file-name (e.g. acetic_45_1ReadsPerGenome.out.tab)
```{r}
# Specify the subfolder where the gene count files are located
count_files_path <- "count_matrix_trimmed"

# List all files in the directory that match the naming pattern
files <- list.files(path = count_files_path, pattern = "ReadsPerGene\\.out\\.tab$", full.names = TRUE)

# Function to read and format each file
read_counts <- function(file) {
  counts <- read.table(file, header = FALSE, skip = 4)  # Skip the first 4 lines if they're headers
  counts <- counts[, c(1, 2)]  # Select GeneID and unstranded counts columns
  
  # Extract sample name from filename (e.g., "acetic_45_1" from "acetic_45_1ReadsPerGene.out.tab")
  sample_name <- sub("ReadsPerGene\\.out\\.tab$", "", basename(file))
  
  colnames(counts) <- c("GeneID", sample_name)
  return(counts)
}

# Load each file and merge them into a single count matrix
count_list <- lapply(files, read_counts)
count_matrix <- Reduce(function(x, y) merge(x, y, by = "GeneID", all = TRUE), count_list)
rownames(count_matrix) <- count_matrix$GeneID
count_matrix <- count_matrix[, -1]  # Remove the GeneID column
# Filter out tRNA genes
filtered_count_matrix <- count_matrix[!rownames(count_matrix) %in% tRNA_gene_ids, ]

# Check dimensions before and after filtering
cat("Original count matrix dimensions:", dim(count_matrix), "\n")
cat("Filtered count matrix dimensions:", dim(filtered_count_matrix), "\n")


```


## Create Sample Metadata
Check the table if the program correctly matched condition and time to the samples.
```{r}
# Define sample names matching the columns of the count matrix
sample_names <- c("trimmed_acetic_45_1_aligned_", "trimmed_acetic_45_2_aligned_", "trimmed_acetic_120_1_aligned_", "trimmed_acetic_120_2_aligned_",
                  "trimmed_acetic_200_1_aligned_", "trimmed_acetic_200_2_aligned_", 
                  "trimmed_control_45_1_aligned_", "trimmed_control_45_2_aligned_", "trimmed_control_120_1_aligned_", "trimmed_control_120_2_aligned_",
                  "trimmed_control_200_1_aligned_", "trimmed_control_200_2_aligned_")

# Define the metadata for each sample
sample_metadata <- data.frame(
    sample = sample_names,
    condition = c(rep("acetic", 6), rep("control", 6)),
    time_point = c(rep("45", 2), rep("120", 2), rep("200", 2), rep("45", 2), rep("120", 2), rep("200", 2)),
    replicate = rep(1:2, 6)  # Repeat 1 and 2 for each combination of condition and time_point
)
rownames(sample_metadata) <- sample_metadata$sample
# Reorder sample_metadata based on the column names of count_matrix
sample_metadata <- sample_metadata[match(colnames(filtered_count_matrix), rownames(sample_metadata)), ]

## Like in the tutorial
sample_metadata$condition <- #relevel(factor(sample_metadata$condition), ref = "control")

# Check the levels of the sample_metadata
sample_metadata$condition <- relevel(factor(sample_metadata$condition), ref = "control")

# Check if meta data is correct
sample_metadata
```
## Results
Create matrix with the count matrix and the meta data. We care about both condition and time point.
```{r}
# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = filtered_count_matrix,
                              colData = sample_metadata,
                              design = ~ condition + time_point)
dds$condition <- relevel(dds$condition, ref = "control")
# Run DESeq2 analysis
dds <- DESeq(dds)


```
## Sanity Checks and Quality Assesment
The PCA-plot does not look perfect, as I would have assumed there is more difference between the different time points but it seems that the control is more different between time points then the acetic treatment. Still the X-Axis explains 79% of the variance so we still see a big difference between treatments.
```{r}
# Perform variance-stabilizing transformation on our count data
vsd <- vst(dds, blind = FALSE)

# Generate a PCA plot with DESeq2's plotPCA function
plotPCA(vsd, intgroup = c("condition", "time_point"))

```
## Heatmap

You can see the same in the heatmap. Acetic treatments are super similar while the control group has a little bit of variation between the two time points.
```{r}
# Calculate distances between samples in our transformed data
sample_dists <- dist(t(assay(vsd)))

# Convert the output to a matrix
sample_dist_matrix <- as.matrix(sample_dists)

# Remove the column names of our matrix
colnames(sample_dist_matrix) <- NULL

# Set the color palette for our heatmap
colours <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)

# Generate a heatmap using the pheatmap package
pheatmap(sample_dist_matrix,
         clustering_distance_rows = sample_dists,
         clustering_distance_cols = sample_dists, 
         col = colours)
```

## Extracting Data from DESeq2 and Filtering
Because of the small difference between time points I decided to only care about acetic vs. control in the downstram analysis.
```{r}
# Display the names of the results calculated by DESeq2
resultsNames(dds)

# Extract the results for our specific comparison
# Replace "condition_treated_vs_control" with the actual result name if different
res <- results(dds, name = "condition_acetic_vs_control") %>% as.data.frame()

# View the first few rows to understand the structure of the results
head(res)
```
## Data Manipulation and Wrangling with dplyr
We can see that we loose 367 genes when dropping rows with NA values. 
```{r}
# Quick overview of the data
glimpse(res)
# Remove rows with NA values in any column
res_no_NA <- res %>% drop_na()

# Check the number of rows remaining after filtering out NA values
glimpse(res_no_NA)

```

## Filter for Significantly Differentially Expressed
```{r}
# Filter to keep only genes with adjusted p-value < 0.05
res_filtered <- res_no_NA %>% filter(padj < 0.05)

# Check the number of rows remaining after this filtering step
glimpse(res_filtered)

# Further filter for biologically relevant genes (log2 fold change ≥ ±1)
res_filtered_final <- res_filtered %>%
  filter(log2FoldChange <= -1 | log2FoldChange >= 1) %>%
  rownames_to_column("gene_id")  # Convert row names to a "gene_id" column for easy handling

# View the first few rows of the final filtered results
head(res_filtered_final)

```

```{r}
# remove rows that contain tRNA genes
#res_filtered_final_notrna <-  res_filtered_final[-c(732:889), ]

# re-number the rows
#rownames(res_filtered_final_notrna) <- 1:nrow(res_filtered_final_notrna)

# view the final filtered data
#view(res_filtered_final_notrna)
```

## Top 10 up- and downregulated genes
Here we could check the gene id's and see whether they match with our hypothesis.
```{r}
# Select the top 10 upregulated genes (highest log2 fold change)
top50_genes <- res_filtered_final %>%
  arrange(desc(log2FoldChange)) %>%
  head(n = 50)

# View the top 10 upregulated genes
top50_genes

# Save top 50 genes as a tab-separated file
write.table(top50_genes, file = "top50_genes_trimmed.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
```
```{r}
# Select the top 10 downregulated genes (lowest log2 fold change)
bot50_genes <- res_filtered_final%>%
  arrange(log2FoldChange) %>%
  head(n = 50)

# View the top 10 downregulated genes
bot50_genes

# Save bop 50 genes as a tab-separated file
write.table(bot50_genes, file = "bot50_genes_trimmed.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

```
## Volcano plot
Volcano plot for acetic vs. control. Problematic with the volcano plot is that some data points are cut up. The Reason for that is that the p-values for certain genes are very very small in comparison to other genes. You could cap the low p-values but I don't know if I like it to "missrepresent" the data.
### Prepare data for volcano plot
```{r}
# Convert the results to a data frame
dat <- as.data.frame(res_filtered_final)

# Label the genes based on log2FoldChange and padj values
labelled_dat <- dat %>%
  mutate(up_down = case_when(
    padj < 0.05 & log2FoldChange > 2 ~ "UP",    # Upregulated genes
    padj < 0.05 & log2FoldChange < -2 ~ "DOWN",  # Downregulated genes
    TRUE ~ "NONE"                                # Not significantly different
  ))
```
## ggplot
```{r}
# Create the volcano plot
labelled_dat %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = up_down)) + 
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # Threshold line for significance
  geom_vline(xintercept = c(-2, 2), linetype = "dashed") +      # Thresholds for fold change
  labs(title = "Volcano Plot of Differential Expression",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value") +
  scale_color_manual(values = c("UP" = "red", "DOWN" = "blue", "NONE" = "gray")) +
  theme_minimal() 

```

## Barplot
We could use the barplot to show that the majority of genes are upregulated. Especially, for the genes with the highest difference between treatments.
### Create new column
```{r}
# Prepare the data for the bar plot
labelled_dat2 <- labelled_dat %>%
  mutate(
    abs_l2fc = abs(log2FoldChange),  # Calculate the absolute value of log2 fold change
    up_down = if_else(log2FoldChange <= 0, "DOWN", "UP")  # Label genes as UP or DOWN
  ) %>% 
  mutate(
    l2fc = case_when(  # Bin genes based on log2 fold change
      abs_l2fc >= 4 ~ ">4",
      abs_l2fc >= 3 ~ "3",
      abs_l2fc >= 2 ~ "2",
      abs_l2fc >= 1 ~ "1",
      TRUE ~ "<1"
    )
  )
```
### Plot the Bar Graph
```{r}
# Initial bar plot
labelled_dat2 %>% 
  ggplot(aes(x = l2fc, fill = up_down)) +
  geom_bar(color = "black") +
  theme_classic() +
  labs(x = "Log2(Fold Change)", y = "# of Genes", fill = "Regulated")

```

### Organize X-Axis

```{r}
# Convert `l2fc` into a factor and specify the desired order
labelled_dat2$l2fc <- factor(labelled_dat2$l2fc, levels = c("<1", "1", "2", "3", ">4"))

```
### Plotting with ordered factors
```{r}
# Final bar plot with ordered x-axis
ggplot(labelled_dat2, aes(x = l2fc, fill = up_down)) +
  geom_bar(color = "black") +
  theme_classic() +
  labs(x = "Log2(Fold Change)", y = "# of Genes", fill = "Regulated") 

```

## Save file as csv
```{r}
write_csv(res_filtered_final, "yeast_results_trimmed.csv")
```

